feat(doc): Add a section about plugin documentation (#136)
kestra-io/kestra.io: docs/plugin-developer-guide/documentation/README.md

---
order: 8
---
# Document your plugin

First, let remember ourself the organization of a plugin project:

> - The Gradle project is called a group of plugins as it contains multiple plugins.

Author: > - The Gradle project can contain several plugins, we call it a group of plugins.

> - The Gradle project is called a group of plugins as it contains multiple plugins.

Reviewer: That sounds good!

- The package in which a plugin is, is called a sub-group of plugins, sometimes there is only one sub-group so the group and the sub-group are the same.

> - Each class is a plugin that provides one of: task, trigger, condition, ...

Author: Yes, a Java class extends a base class that is the `Task` class, the `Trigger` class, the `Condition` task, ...

The documentation will be used on the Kestra website and on the Kestra UI.

We provides a way to document each level of a plugin project.

## Document the plugin group

### Manifest attributes

Kestra uses custom manifest attributes to provides group top level documentation.

The following manifest attributes are used to document the group of plugins:

> - `X-Kestra-Title`: we usually use the Gradle `project.name` property.

Author: Those are defaults set by the plugin template repository. I can replace each by:

> by default, the Gradle `project.name` property is used.

> - `X-Kestra-Title`: we usually use the Gradle `project.name` property.

Reviewer: Sounds good.

- `X-Kestra-Group`: we usually use the Gradle `group.id` property with an additional group name.
- `X-Kestra-Description`: we usually use the Gradle `project.description` property.
- `X-Kestra-Version`: we usually use the Gradle `project.version` property.

If you follow the plugin structure of the GitHub template you should have something like that:

> ```ruby

Reviewer 2: ```suggestion
```groovy
```

group "io.kestra.plugin"
description 'Google Cloud Platform (GCP) plugins for Kestra.'

// [...]

jar {
    manifest {
        attributes(
                "X-Kestra-Title": project.name,
                "X-Kestra-Group": project.group + ".gcp",
                "X-Kestra-Description": project.description,
                "X-Kestra-Version": project.version
        )
    }
}
```

> As you can see, the most important documentation attribute is the `description` that should be a short sentence that describe the group of plugins.

Author: I think we should keep 'group of plugins' here

> As you can see, the most important documentation attribute is the `description` that should be a short sentence that describe the group of plugins.

Reviewer: Okay, sounds good.

You can add multiple examples if needed.

### Document the plugin properties

In a plugin, all properties must be annotated by `io.kestra.core.models.annotations.PluginProperty` and should provide documentation via the `io.swagger.v3.oas.annotations.media.Schema` annotation and validation rules via `javax.validation.constraints.*`.

For validation rules, please check the [Properties validation](../runnable-task/#properties-validation) documentation page.

The `@PluginProperty` annotation contains two attributes:

- `dynamic`: set it to true if the property will be rendered dynamically, see [Render Variable](..//runnable-task/#render-variables).
- `additionalProperties`: you can use it to denote the sub-type of the property, for example when using a `Map<String, Message>` you can set it to `Message.class`.

The Swagger `@Schema` annotation contains a lot of attributes that can be used to document the plugin properties, the most useful are:

- `title`: short description of a property.
- `description`: long description of a property.

> - `anyOf`: allows to specify which types are allowed when the type of the property can be extended (interface, abstract classes, ...). This should be set when the type of the property is `Object`.

Author: I had a hard time explaining this ...
I think the best would be to describe more precisely how to use it

> specifying which sub-types are allowed when the property type is an interface, an abstract class, or a class inside a hierarchy of classes.

The `@Schema` and `@PluginProperty` annotations can be used on fields, methods or classes.

A lot of tasks can take input from multiple sources on the same property, they usually have a single `from` property that can be a string that represent a file in the Kestra Storage, a single object or a list of objects. To document such property you can use `anyOf` this way:

```java
@PluginProperty(dynamic = true)
@NotNull
@Schema(
    title = "The source of the published data.",
    description = "Can be an internal storage URI, a list of Pub/Sub messages, or a single Pub/Sub message.",
    anyOf = {String.class, Message[].class, Message.class}
)
private Object from;
```

> Note that, due to limitation on the way JSON schema works, you cannot add `@Schema` both on a Java enum type and the plugin property that uses this type. We advise to avoid using `@Schema` on enumerations.

Reviewer: ```suggestion
**Note:** Due to limitations on how JSON schema works, you cannot add `@Schema` on a Java enum type and the plugin property that uses this type. We advise avoiding using `@Schema` on enumerations.
```

Author: I'm pretty sure we have a format to display notes as a special visual block, I'll check and update the formating.

### Document the plugin outputs

> Outputs should be documented with the `io.swagger.v3.oas.annotations.media.Schema` annotation the same way as plugin properties, please read the section above for more information.

Reviewer: ```suggestion
Outputs should be documented with the `io.swagger.v3.oas.annotations.media.Schema` annotation the same way as plugin properties. Please read the section above for more information.
```

> Only this annotation should be used, never use `@PluginProperty` on an output.

Reviewer: ```suggestion
Only use the annotation mentioned above. Never use `@PluginProperty` on an output.
```
